samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -0,0 +1,172 @@
+    this(processorId, config, customMetricsReporters, (Object) asyncStreamTaskFactory);
+  }
+
+  /**
+   * Same as {@link #StreamProcessor(int, Config, Map, AsyncStreamTaskFactory)}, except task instances are created
+   * using the provided {@link StreamTaskFactory}.
+   */
+  public StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters,
+                         StreamTaskFactory streamTaskFactory) {
+    this(processorId, config, customMetricsReporters, (Object) streamTaskFactory);
+  }
+
+  /**
+   * Same as {@link #StreamProcessor(int, Config, Map, AsyncStreamTaskFactory)}, except task instances are created
+   * using the "task.class" configuration instead of a task factory.
+   */
+  public StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters) {
+    this(processorId, config, customMetricsReporters, (Object) null);
+  }
+
+  private StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters,
+                          Object taskFactory) {
+    this.processorId = processorId;
+
+    Map<String, String> updatedConfigMap = new HashMap<>();
+    updatedConfigMap.putAll(config);
+    updatedConfigMap.put(PROCESSOR_ID, String.valueOf(this.processorId));
+    Config updatedConfig = new MapConfig(updatedConfigMap);
+
+
+    this.containerController = new SamzaContainerController(
+        taskFactory,
+        new TaskConfigJava(updatedConfig).getShutdownMs(),
+        String.valueOf(processorId),
+        customMetricsReporters);
+
+    this.jobCoordinator = Util.
+        <JobCoordinatorFactory>getObj(
+            new JobCoordinatorConfig(updatedConfig)
+                .getJobCoordinatorFactoryClassName())
+        .getJobCoordinator(processorId, updatedConfig, this.containerController);
+  }
+
+  /**
+   * StreamProcessor Lifecycle: start()
+   * <ul>
+   * <li>Starts the JobCoordinator and fetches the JobModel</li>
+   * <li>Starts the container using ContainerModel based on the processorId </li>
+   * </ul>
+   * When start() returns, it only guarantees that the container is initialized and submitted by the controller to
+   * execute
+   */
+  public void start() {
+    jobCoordinator.start();
+  }
+
+  /**
+   * Method that allows the user to wait for a specified amount of time for the container to initialize and start
+   * processing messages
+   *
+   * @param timeoutMs Maximum time to wait, in milliseconds
+   * @return {@code true}, if the container started within the specified wait time and {@code false} if the waiting time
+   * elapsed
+   * @throws InterruptedException if the current thread is interrupted while waiting for container to start-up
+   */
+  public boolean awaitStart(long timeoutMs) throws InterruptedException {
+    return containerController.awaitStart(timeoutMs); // TODO: Should awaitStart be part of the JC interface, instead of directly using container controller
```

Comment:
Oh, I mean this: instead of containerController.awaitStart(timeoutMs), we do jobCoordinator.awaitStart(), as the comment suggested (it make a lot of sense to me since the controller is started and stopped in coordinator.). That way containerController can be directly created inside JobCoordinator, and we don't need to have it inside StreamProcessor and pass it in to the jobCoordinatorFactory. And the containerController can be package private too.